spiffe-helper.go

```diff
@@ -13,63 +11,93 @@ import (
 	"strings"
 	"syscall"
 	"time"
-
-	"github.com/spiffe/spire/proto/api/workload"
+	"crypto/x509"
+	"github.com/spiffe/spire/api/workload"
+	"net"
+	proto "github.com/spiffe/spire/proto/api/workload"
 )
 
-// Sidecar is the component that consumes Workload API and renews certs
-type Sidecar struct {
-	config                *SidecarConfig
-	workloadClient        workload.WorkloadClient
-	workloadClientContext context.Context
-	processRunning        bool
-	process               *os.Process
+type Sidecar interface {
+	RunDaemon()
+}
+
+// sidecar is the component that consumes the Workload API and renews certs
+// implements the interface Sidecar
+type sidecar struct {
+	config            *SidecarConfig
+	processRunning    bool
+	process           *os.Process
+	workloadAPIClient workload.X509Client
 }
 
+// default timeout Duration for the workloadAPI client when the timeout
+// is not configured in the .conf file
+const timeout = time.Duration(5 * time.Second)
+
 // NewSidecar creates a new sidecar
-func NewSidecar(workloadClientContext context.Context, config *SidecarConfig, workloadClient workload.WorkloadClient) *Sidecar {
-	return &Sidecar{
-		workloadClientContext: workloadClientContext,
-		config:                config,
-		workloadClient:        workloadClient,
+func NewSidecar(config *SidecarConfig) *sidecar {
+	return &sidecar{
+		config:            config,
+		workloadAPIClient: newWorkloadAPIClient(config.AgentAddress, getTimeout(config)),
 	}
 }
 
 // RunDaemon starts the main loop
-func (s *Sidecar) RunDaemon() error {
+// Starts the workload API client to listen for new SVID updates
+// When a new SVID is received on the updateChan, the SVID certificates
+// are stored in disk and a restart signal is sent to the proxy's process
+func (s *sidecar) RunDaemon() error {
 	// Create channel for interrupt signal
 	interrupt := make(chan os.Signal, 1)
+	errorChan := make(chan error)
 	signal.Notify(interrupt, syscall.SIGINT, syscall.SIGTERM)
 
-	// Main loop
-	for {
-		// Fetch and dump certificates
-		ttl, err := s.dumpBundles()
-		if err != nil {
-			return err
-		}
-		err = s.signalProcess()
+	updateChan := s.workloadAPIClient.UpdateChan()
+
+	//start the workloadAPIClient
+	go func() {
+		err := s.workloadAPIClient.Start()
```

Comment:
seems like Start() returning for any reason would be problematic.

Reply:
workloadAPIClient.Start() always returns Fatal error, because it couldn't connect with the Workload API after several retries

Reply:
The workloadAPI client retries failed operations, so I think when it returns an error it could considered fatal